File: datasets/rs/HumanEval_115_max_fill.rs
/// You are given a rectangular grid of wells. Each row represents a single well,
/// and each 1 in a row represents a single unit of water.
/// Each well has a corresponding bucket that can be used to extract water from it, 
/// and all buckets have the same capacity.
/// Your task is to use the buckets to empty the wells.
/// Output the number of times you need to lower the buckets.
/// Example 1:
/// Input: 
/// grid : [[0,0,1,0], [0,1,0,0], [1,1,1,1]]
/// bucket_capacity : 1
/// Output: 6
/// Example 2:
/// Input: 
/// grid : [[0,0,1,1], [0,0,0,0], [1,1,1,1], [0,1,1,1]]
/// bucket_capacity : 2
/// Output: 5
/// Example 3:
/// Input: 
/// grid : [[0,0,0], [0,0,0]]
/// bucket_capacity : 5
/// Output: 0
/// Constraints:
/// * all wells have the same length
/// * 1 <= grid.length <= 10^2
/// * 1 <= grid[:,1].length <= 10^2
/// * grid[i][j] -> 0 | 1
/// * 1 <= capacity <= 10
fn max_fill(grid: Vec<Vec<isize>>, capacity: isize) -> isize {
    let mut grid = grid;
    let mut max_fill = 0;
    for i in 0..grid.len() {
        for j in 0..grid[i].len() {
            if grid[i][j] == 1 {
                let mut fill = 0;
                let mut fill_grid = grid.clone();
                let mut queue = VecDeque::new();
                queue.push_back((i, j));
                while !queue.is_empty() {
                    let (x, y) = queue.pop_front().unwrap();
                    fill += 1;
                    fill_grid[x][y] = 0;
                    if x > 0 && fill_grid[x - 1][y] == 1 {
                        queue.push_back((x - 1, y));
                    }
                    if x < fill_grid.len() - 1 && fill_grid[x + 1][y] == 1 {
                        queue.push_back((x + 1, y));
                    }
                    if y > 0 && fill_grid[x][y - 1] == 1 {
                        queue.push_back((x, y - 1));
                    }
                    if y < fill_grid[x].len() - 1 && fill_grid[x][y + 1] == 1 {
                        queue.push_back((x, y + 1));
                    }
                }
                if fill > max_fill {
                    max_fill = fill;
                }
            }
        }
    }
    max_fill

}

fn main() {
    let candidate = max_fill;
    assert_eq!(candidate(vec![vec![0, 0, 1, 0], vec![0, 1, 0, 0], vec![1, 1, 1, 1]], 1), 6);
    assert_eq!(candidate(vec![vec![0, 0, 1, 1], vec![0, 0, 0, 0], vec![1, 1, 1, 1], vec![0, 1, 1, 1]], 2), 5);
    assert_eq!(candidate(vec![vec![0, 0, 0], vec![0, 0, 0]], 5), 0);
    assert_eq!(candidate(vec![vec![1, 1, 1, 1], vec![1, 1, 1, 1]], 2), 4);
    assert_eq!(candidate(vec![vec![1, 1, 1, 1], vec![1, 1, 1, 1]], 9), 2);
}

File: datasets/java/HumanEval_115_max_fill.java
import java.util.*;
import java.lang.reflect.*;
import org.javatuples.*;
class Problem {
private static <T> T getDefaultValue(Class<T> clazz) {
return (T) Array.get(Array.newInstance(clazz, 1), 0);
}    // You are given a rectangular grid of wells. Each row represents a single well,
    // and each 1 in a row represents a single unit of water.
    // Each well has a corresponding bucket that can be used to extract water from it, 
    // and all buckets have the same capacity.
    // Your task is to use the buckets to empty the wells.
    // Output the number of times you need to lower the buckets.
    // Example 1:
    // Input: 
    // grid : [[0,0,1,0], [0,1,0,0], [1,1,1,1]]
    // bucket_capacity : 1
    // Output: 6
    // Example 2:
    // Input: 
    // grid : [[0,0,1,1], [0,0,0,0], [1,1,1,1], [0,1,1,1]]
    // bucket_capacity : 2
    // Output: 5
    // Example 3:
    // Input: 
    // grid : [[0,0,0], [0,0,0]]
    // bucket_capacity : 5
    // Output: 0
    // Constraints:
    // * all wells have the same length
    // * 1 <= grid.length <= 10^2
    // * 1 <= grid[:,1].length <= 10^2
    // * grid[i][j] -> 0 | 1
    // * 1 <= capacity <= 10
    public static long max_fill(ArrayList<ArrayList<Long>> grid, long capacity) {
        int n = grid.size();
        int m = grid.get(0).size();
        long[][] dp = new long[n][m];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                dp[i][j] = grid.get(i).get(j);
            }
        }
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (i == 0 && j == 0) continue;
                if (i == 0) {
                    dp[i][j] = Math.min(dp[i][j], dp[i][j - 1]);
                } else if (j == 0) {
                    dp[i][j] = Math.min(dp[i][j], dp[i - 1][j]);
                } else {
                    dp[i][j] = Math.min(dp[i][j], Math.max(dp[i - 1][j], dp[i][j - 1]));
                }
                dp[i][j] = Math.min(dp[i][j], capacity);
            }
        }
        return dp[n - 1][m - 1];
    }
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        int m = scanner.nextInt();
        long capacity = scanner.nextLong();
        ArrayList<ArrayList<Long>> grid = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            ArrayList<Long> row = new ArrayList<>();
            for (int j = 0; j < m; j++) {
                row.add(scanner.nextLong());
            }
            grid.add(row);
        }
        System.out.println(max_fill(grid, capacity));
    }
}

    }
    public static void main(String[] args) {
    assert(max_fill((new ArrayList<ArrayList<Long>>(Arrays.asList((ArrayList<Long>)new ArrayList<Long>(Arrays.asList((long)0l, (long)0l, (long)1l, (long)0l)), (ArrayList<Long>)new ArrayList<Long>(Arrays.asList((long)0l, (long)1l, (long)0l, (long)0l)), (ArrayList<Long>)new ArrayList<Long>(Arrays.asList((long)1l, (long)1l, (long)1l, (long)1l))))), (1l)) == (6l));
    assert(max_fill((new ArrayList<ArrayList<Long>>(Arrays.asList((ArrayList<Long>)new ArrayList<Long>(Arrays.asList((long)0l, (long)0l, (long)1l, (long)1l)), (ArrayList<Long>)new ArrayList<Long>(Arrays.asList((long)0l, (long)0l, (long)0l, (long)0l)), (ArrayList<Long>)new ArrayList<Long>(Arrays.asList((long)1l, (long)1l, (long)1l, (long)1l)), (ArrayList<Long>)new ArrayList<Long>(Arrays.asList((long)0l, (long)1l, (long)1l, (long)1l))))), (2l)) == (5l));
    assert(max_fill((new ArrayList<ArrayList<Long>>(Arrays.asList((ArrayList<Long>)new ArrayList<Long>(Arrays.asList((long)0l, (long)0l, (long)0l)), (ArrayList<Long>)new ArrayList<Long>(Arrays.asList((long)0l, (long)0l, (long)0l))))), (5l)) == (0l));
    assert(max_fill((new ArrayList<ArrayList<Long>>(Arrays.asList((ArrayList<Long>)new ArrayList<Long>(Arrays.asList((long)1l, (long)1l, (long)1l, (long)1l)), (ArrayList<Long>)new ArrayList<Long>(Arrays.asList((long)1l, (long)1l, (long)1l, (long)1l))))), (2l)) == (4l));
    assert(max_fill((new ArrayList<ArrayList<Long>>(Arrays.asList((ArrayList<Long>)new ArrayList<Long>(Arrays.asList((long)1l, (long)1l, (long)1l, (long)1l)), (ArrayList<Long>)new ArrayList<Long>(Arrays.asList((long)1l, (long)1l, (long)1l, (long)1l))))), (9l)) == (2l));
    }

}
